thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/TriggerConditionGrouper.java

```diff
@@ -0,0 +1,205 @@
+    List<MergedAnomalyResultDTO> anomalies = mergeAndSortAnomalies(anomalyListA, anomalyListB);
+    if (anomalies.isEmpty()) {
+      return anomalies;
+    }
+
+    for (int i = 0; i < anomalies.size(); i++) {
+      for (int j = i + 1; j < anomalies.size(); j++) {
+        // Check for overlap and output it
+        if (anomalies.get(j).getStartTime() <= anomalies.get(i).getEndTime()) {
+          MergedAnomalyResultDTO currentAnomaly = makeParentEntityAnomaly(anomalies.get(i));
+          currentAnomaly.setEndTime(Math.min(currentAnomaly.getEndTime(), anomalies.get(j). getEndTime()));
+          currentAnomaly.setStartTime(anomalies.get(j). getStartTime());
+          setEntityChildMapping(currentAnomaly, anomalies.get(j));
+
+          groupedAnomalies.add(currentAnomaly);
+        } else {
+          break;
+        }
+      }
+    }
+
+    return new ArrayList<>(groupedAnomalies);
+  }
+
+  /**
+   * Group based on 'OR' criteria - Entity has anomaly if either sub-entity A or B have anomalies.
+   * This means we find the total anomaly coverage.
+   *
+   * Since the anomalies from the respective entities/metrics are merged
+   * before calling the grouper, we do not have to deal with overlapping
+   * anomalies within an entity/metric
+   *
+   * Sort anomalies by start time and incrementally merge anomalies
+   */
+  private List<MergedAnomalyResultDTO> orGrouping(
+      List<MergedAnomalyResultDTO> anomalyListA, List<MergedAnomalyResultDTO> anomalyListB) {
+    Set<MergedAnomalyResultDTO> groupedAnomalies = new HashSet<>();
+    List<MergedAnomalyResultDTO> anomalies = mergeAndSortAnomalies(anomalyListA, anomalyListB);
+    if (anomalies.isEmpty()) {
+      return anomalies;
+    }
+
+    MergedAnomalyResultDTO currentAnomaly = makeParentEntityAnomaly(anomalies.get(0));
+    for (int i = 1; i < anomalies.size();  i++) {
+      if (anomalies.get(i).getStartTime() <= currentAnomaly.getEndTime()) {
+        // Partial or full overlap
+        currentAnomaly.setEndTime(Math.max(anomalies.get(i).getEndTime(), currentAnomaly.getEndTime()));
+        setEntityChildMapping(currentAnomaly, anomalies.get(i));
+      }  else {
+        // No overlap
+        groupedAnomalies.add(currentAnomaly);
+        currentAnomaly = makeParentEntityAnomaly(anomalies.get(i));
+      }
+    }
+    groupedAnomalies.add(currentAnomaly);
+
+    return new ArrayList<>(groupedAnomalies);
+  }
+
+  /**
+   * Groups the anomalies based on the parsed operator tree
+   */
+  private List<MergedAnomalyResultDTO> groupAnomaliesByOperator(Map<String, Object> operatorNode, List<MergedAnomalyResultDTO> anomalies) {
+    Preconditions.checkNotNull(operatorNode);
+
+    // Base condition - If reached leaf node, then return the anomalies corresponding to the entity/metric
+    String value = MapUtils.getString(operatorNode, "value");
+    if (value != null) {
+      List<MergedAnomalyResultDTO> filteredAnomalies = new ArrayList<>();
```

Comment:
consider using the filter in the steam API.
```
filteredAnomalies = anomalies.steam().filter(...).collect(Collector.toList());
```

Reply:
updated